### PR TITLE
Allowed Record to be called with no arguments

### PR DESCRIPTION
**Pull request by alexarchambault**
Allows the following to compile:
```scala
val r = Record()
```

### CODE REVIEW

**Comment by milessabin** [comment]
This is basically another name for `HNil`. Could you give me a rationale for adding it?

**Comment by alexarchambault** [comment]
"Makes the intent clearer" ?

File: core/src/main/scala/shapeless/records.scala [hunk 1]
```diff
@@ -98,6 +99,13 @@ class RecordMacros(val c: whitebox.Context) {
   val SymTpe = typeOf[scala.Symbol]
   val atatTpe = typeOf[tag.@@[_,_]].typeConstructor
 
+  def mkRecordEmptyImpl(method: Tree)(rec: Tree*): Tree = {
+    if(rec.nonEmpty)
+      c.abort(c.enclosingPosition, "this method must be called with named arguments")
```
Comment (milessabin) [inline core/src/main/scala/shapeless/records.scala]: s/named/no/ ?

File: core/src/main/scala/shapeless/records.scala [hunk 2]
```diff
@@ -98,6 +99,13 @@ class RecordMacros(val c: whitebox.Context) {
...
+    
+    mkRecordNamedImpl(method)(rec: _*)
```
Comment (milessabin) [inline core/src/main/scala/shapeless/records.scala]: q"_root_.shapeless.HNil"?

File: core/src/main/scala/shapeless/records.scala [hunk 3]
```diff
@@ -56,6 +56,7 @@ object record {
    * }}}
    */
   object Record extends Dynamic {
+    def applyDynamic(method: String)(rec: Any*): HList = macro RecordMacros.mkRecordEmptyImpl
```
Comment (milessabin) [inline core/src/main/scala/shapeless/records.scala]: Presumably a vanilla `apply()` would preempt the `applyDynamicNamed`?

**Comment by milessabin** [comment]
Fair enough. But I'm slightly nervous that this might give the impression that shapeless records are more first-class than they are.

Happy to merge, modulo the comments I've just made above.

File: core/src/main/scala/shapeless/records.scala (hunk 3 above)
Comment (alexarchambault) [inline core/src/main/scala/shapeless/records.scala]: It does, indeed. Errors like:
```scala
[error] /home/alexandre/projects/shapeless/core/src/test/scala/shapeless/product.scala:132: too many arguments for method apply: ()shapeless.HNil.type in object Record
[error]       val expectedFooL = Record(i = 1, s = "b")
[error]                                ^
```

File: core/src/main/scala/shapeless/records.scala (hunk 1 above)
Comment (alexarchambault) [inline core/src/main/scala/shapeless/records.scala]: This error message appears when `Record` is called like
```scala
Record(2, "a")
```

So this error message ("must be called with *named* arguments") looked like the right one to me.

File: core/src/main/scala/shapeless/records.scala (hunk 2 above)
Comment (alexarchambault) [inline core/src/main/scala/shapeless/records.scala]: Better this way, yes, I'll update the PR.

**Comment by alexarchambault** [comment]
Updated the PR. What do you mean by not giving the impression that records are more first-class than they are? Aren't they a useful ~collection (which shapeless makes all the more easy to use)?

**Comment by milessabin** [comment]
Really it's just a vague worry that people might forget that they are just HLists. No big deal.

File: core/src/main/scala/shapeless/records.scala (hunk 1 above)
Comment (milessabin) [inline core/src/main/scala/shapeless/records.scala]: Makes sense.

**Comment by milessabin** [comment]
In any case, LGTM ... many thanks ... merging :-)

**Pull request merged**